substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java

```diff
@@ -0,0 +1,158 @@
+
+    private static final int TRACE_ID_SHIFT = 16;
+
+    // Epoch stuff
+    private static final long USED_BIT = 1;
+    private static final int EPOCH_1_SHIFT = 0;
+    private static final int EPOCH_2_SHIFT = 1;
+    private static final long USED_EPOCH_1_BIT = USED_BIT << EPOCH_1_SHIFT;
+    private static final long USED_EPOCH_2_BIT = USED_BIT << EPOCH_2_SHIFT;
+
+    private static final long JDK_JFR_EVENT_SUBKLASS = 16;
+    private static final long JDK_JFR_EVENT_KLASS = 32;
+    private static final long EVENT_HOST_KLASS = 64;
+
+    @Fold
+    static JfrTraceIdMap getTraceIdMap() {
+        return ImageSingletons.lookup(JfrTraceIdMap.class);
+    }
+
+    public static void tag(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        map.setId(clazz, (id & ~0xff) | (bits & 0xff));
+    }
+
+    public static boolean predicate(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        return (id & bits) != 0;
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static void setUsedThisEpoch(Class<?> clazz) {
+        tag(clazz, JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static boolean isUsedThisEpoch(Class<?> clazz) {
+        return predicate(clazz, TRANSIENT_BIT | JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static long getTraceIdRaw(Class<?> clazz) {
+        return getTraceIdMap().getId(clazz);
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static long getTraceId(Class<?> clazz) {
+        long traceid = getTraceIdRaw(clazz);
+        return traceid >>> TRACE_ID_SHIFT;
+    }
+
+    public static long load(Class<?> clazz) {
+        return JfrTraceIdLoadBarrier.load(clazz);
+    }
+
+    private static void tagAsJdkJfrEvent(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        map.setId(index, id | JDK_JFR_EVENT_KLASS);
+    }
+
+    private static void tagAsJdkJfrEventSub(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        map.setId(index, id | JDK_JFR_EVENT_SUBKLASS);
+    }
+
+    private static boolean isEventClass(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        return (id & (JDK_JFR_EVENT_KLASS | JDK_JFR_EVENT_SUBKLASS)) != 0;
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public static void assign(Class<?> clazz, int index) {
+        assert clazz != null;
+        index = index + 1; // Off-set by one for error-catcher
```

Comment:
Should be detected by `mx checkstyle`

Reply:
mx checkstyle is clean, and I'm not sure what it should find there?

Reply:
We disallow assigning a new value to a method argument. If it is not detected by checkstyle, then this should be detected by something else (probably a warning when building the Java sources).